File: python/leetcodehard/remove_invalid_parentheses.py
"""
https://leetcode.com/problems/remove-invalid-parentheses/

status - completed
"""

from collections import deque

from base_logger.logging_event import create_logger


class Solution:

    log = create_logger(__name__)

    def removeInvalidParentheses(self, s):
        def is_valid(candidate):
            balance = 0
            for char in candidate:
                if char == "(":
                    balance += 1
                elif char == ")":
                    balance -= 1
                    if balance < 0:
                        return False
            return balance == 0

        visited = {s}
        queue = deque([s])
        results = []
        found = False

        while queue:
            current = queue.popleft()

            if is_valid(current):
                results.append(current)
                found = True

            if found:
                continue

            for i in range(len(current)):
                if current[i] not in ("(", ")"):
                    continue
                candidate = current[:i] + current[i + 1:]
                if candidate not in visited:
                    visited.add(candidate)
                    queue.append(candidate)

        return results if results else [""]


if __name__ == "__main__":
    test_input_1 = "()())()"
    output = Solution().removeInvalidParentheses(test_input_1)
    Solution().log.info("Output is %s", output)

    test_input_2 = "(a)())()"
    output = Solution().removeInvalidParentheses(test_input_2)
    Solution().log.info("Output is %s", output)

File: typescript/leetcodehard/removeInvalidParentheses.ts
/*
https://leetcode.com/problems/remove-invalid-parentheses/

status - completed
*/

import { logp } from "../utils/logger";

function isValid(s: string): boolean {
  let balance = 0;
  for (const ch of s) {
    if (ch === "(") balance++;
    else if (ch === ")") balance--;
    if (balance < 0) return false;
  }
  return balance === 0;
}

function removeInvalidParentheses(s: string): string[] {
  const visited = new Set<string>([s]);
  let queue = [s];
  const result: string[] = [];
  let found = false;

  while (queue.length > 0 && !found) {
    const nextQueue: string[] = [];

    for (const candidate of queue) {
      if (isValid(candidate)) {
        result.push(candidate);
        found = true;
        continue;
      }

      for (let i = 0; i < candidate.length; i++) {
        if (candidate[i] !== "(" && candidate[i] !== ")") continue;
        const next = candidate.slice(0, i) + candidate.slice(i + 1);
        if (!visited.has(next)) {
          visited.add(next);
          nextQueue.push(next);
        }
      }
    }

    queue = nextQueue;
  }

  return result;
}

logp(
  `input "()())" expected [()(), (())] (order may vary) actual ${JSON.stringify(
    removeInvalidParentheses("()())").sort(),
  )}`,
);
logp(`input "(a)())()" expected [(a)()(), (a())()] (order may vary) actual ${JSON.stringify(
  removeInvalidParentheses("(a)())()").sort(),
)}`);
logp(`input ")(" expected [""] actual ${JSON.stringify(removeInvalidParentheses(")("))}`);
